// @blockcerts/cert-verifier-js/lib/parsers/parseV3.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = parseV3;

require("core-js/modules/es.promise.js");

var _ldsMerkleProof = require("@vaultie/lds-merkle-proof-2019");

var _constants = require("../constants");

var _domain = _interopRequireDefault(require("../domain"));

function parseSignature(signature) {
  const base58Decoder = new _ldsMerkleProof.Decoder(signature.proofValue);
  return base58Decoder.decode();
}

function getRecipientFullName(certificateJson) {
  const {
    credentialSubject
  } = certificateJson;
  return credentialSubject.name || '';
}

async function parseV3(certificateJson) {
  const receipt = parseSignature(certificateJson.proof);
  const {
    issuer: issuerProfileUrl,
    metadataJson,
    metadata,
    issuanceDate,
    id,
    expirationDate
  } = certificateJson;
  const certificateMetadata = metadata || metadataJson;
  const issuer = await _domain.default.verifier.getIssuerProfile(issuerProfileUrl);
  return {
    chain: _domain.default.certificates.getChain('', receipt),
    expires: expirationDate,
    issuedOn: issuanceDate,
    id,
    issuer,
    metadataJson: certificateMetadata,
    proof: certificateJson.proof,
    receipt,
    recipientFullName: getRecipientFullName(certificateJson),
    recordLink: id,
    // TODO: more dynamic set up of V3
    version: _constants.CERTIFICATE_VERSIONS.V3_0_alpha
  };
}
