// @blockcerts/cert-verifier-js-v1-legacy/lib/index.d.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "BLOCKCHAINS", {
  enumerable: true,
  get: function () {
    return _constants.BLOCKCHAINS;
  }
});
Object.defineProperty(exports, "STEPS", {
  enumerable: true,
  get: function () {
    return _constants.STEPS;
  }
});
Object.defineProperty(exports, "SUB_STEPS", {
  enumerable: true,
  get: function () {
    return _constants.SUB_STEPS;
  }
});
Object.defineProperty(exports, "CERTIFICATE_VERSIONS", {
  enumerable: true,
  get: function () {
    return _constants.CERTIFICATE_VERSIONS;
  }
});
Object.defineProperty(exports, "VERIFICATION_STATUSES", {
  enumerable: true,
  get: function () {
    return _constants.VERIFICATION_STATUSES;
  }
});
Object.defineProperty(exports, "getSupportedLanguages", {
  enumerable: true,
  get: function () {
    return _useCases.getSupportedLanguages;
  }
});
Object.defineProperty(exports, "SignatureImage", {
  enumerable: true,
  get: function () {
    return _models.SignatureImage;
  }
});
exports.TRANSACTION_APIS = exports.SupportedChains = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _constants = require("./constants");

var _useCases = require("./domain/i18n/useCases");

var _models = require("./models");

// TODO: improve definition and export from this file
let SupportedChains;
exports.SupportedChains = SupportedChains;

(function (SupportedChains) {
  SupportedChains["Bitcoin"] = "bitcoin";
  SupportedChains["Ethmain"] = "ethmain";
  SupportedChains["Ethropst"] = "ethropst";
  SupportedChains["Ethrinkeby"] = "ethrinkeby";
  SupportedChains["Mocknet"] = "mocknet";
  SupportedChains["Regtest"] = "regtest";
  SupportedChains["Testnet"] = "testnet";
})(SupportedChains || (exports.SupportedChains = SupportedChains = {}));

let TRANSACTION_APIS;
exports.TRANSACTION_APIS = TRANSACTION_APIS;

(function (TRANSACTION_APIS) {
  TRANSACTION_APIS["blockcypher"] = "blockcypher";
  TRANSACTION_APIS["blockstream"] = "blockstream";
  TRANSACTION_APIS["etherscan"] = "etherscan";
})(TRANSACTION_APIS || (exports.TRANSACTION_APIS = TRANSACTION_APIS = {}));
