// @blockcerts/cert-verifier-js-v1-legacy/lib/domain/certificates/useCases/getVerificationMap.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = getVerificationMap;

var _constants = require("../../../constants");

var _chains = _interopRequireDefault(require("../../chains"));

var _useCases = require("../../i18n/useCases");

var _object = require("../../../helpers/object");

const versionVerificationMap = {
  [_constants.NETWORKS.mainnet]: [_constants.SUB_STEPS.getTransactionId, _constants.SUB_STEPS.computeLocalHash, _constants.SUB_STEPS.fetchRemoteHash, _constants.SUB_STEPS.getIssuerProfile, _constants.SUB_STEPS.parseIssuerKeys, _constants.SUB_STEPS.compareHashes, _constants.SUB_STEPS.checkMerkleRoot, _constants.SUB_STEPS.checkReceipt, _constants.SUB_STEPS.checkRevokedStatus, _constants.SUB_STEPS.checkAuthenticity, _constants.SUB_STEPS.checkExpiresDate],
  [_constants.NETWORKS.testnet]: [_constants.SUB_STEPS.computeLocalHash, _constants.SUB_STEPS.compareHashes, _constants.SUB_STEPS.checkReceipt, _constants.SUB_STEPS.checkExpiresDate]
};
/**
 * stepsObjectToArray
 *
 * Turn an object with steps as properties to an array
 *
 * @param stepsObject
 * @returns {{code: string}[]}
 */

function stepsObjectToArray(stepsObject) {
  return Object.keys(stepsObject).map(stepCode => {
    return { ...stepsObject[stepCode],
      code: stepCode,
      label: (0, _useCases.getText)('steps', `${stepCode}Label`),
      labelPending: (0, _useCases.getText)('steps', `${stepCode}LabelPending`)
    };
  });
}
/**
 * setSubStepsToSteps
 *
 * Takes an array of sub-steps and set them to their proper parent step
 *
 * @param subSteps
 * @returns {any}
 */


function setSubStepsToSteps(subSteps) {
  const steps = (0, _object.deepCopy)(_constants.STEPS.language);
  subSteps.forEach(subStep => steps[subStep.parentStep].subSteps.push(subStep));
  return steps;
}
/**
 * getFullStepsFromSubSteps
 *
 * Builds a full steps array (with subSteps property) from an array of sub-steps
 *
 * @param subStepMap
 * @returns {Array}
 */


function getFullStepsFromSubSteps(subStepMap) {
  const subSteps = subStepMap.map(stepCode => {
    const subStep = Object.assign({}, _constants.SUB_STEPS.language[stepCode]);
    return { ...subStep,
      label: (0, _useCases.getText)('subSteps', `${stepCode}Label`),
      labelPending: (0, _useCases.getText)('subSteps', `${stepCode}LabelPending`)
    };
  });
  const steps = setSubStepsToSteps(subSteps);
  return stepsObjectToArray(steps);
}
/**
 * getVerificationMap
 *
 * Get verification map from the chain
 *
 * @param chain
 * @returns {Array}
 */


function getVerificationMap(chain) {
  if (!chain) {
    return [];
  }

  const network = _chains.default.isMockChain(chain) ? _constants.NETWORKS.testnet : _constants.NETWORKS.mainnet;
  const verificationMap = Object.assign(versionVerificationMap);
  return getFullStepsFromSubSteps(verificationMap[network]);
}
