// @blockcerts/cert-verifier-js/lib/constants/verificationSteps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.language = exports.VerificationSteps = exports.final = void 0;
const final = 'final';
exports.final = final;
let VerificationSteps;
exports.VerificationSteps = VerificationSteps;

(function (VerificationSteps) {
  VerificationSteps["formatValidation"] = "formatValidation";
  VerificationSteps["hashComparison"] = "hashComparison";
  VerificationSteps["statusCheck"] = "statusCheck";
  VerificationSteps["final"] = "final";
})(VerificationSteps || (exports.VerificationSteps = VerificationSteps = {}));

const language = {
  [VerificationSteps.formatValidation]: {
    label: 'Format validation',
    labelPending: 'Validating format',
    subSteps: []
  },
  [VerificationSteps.hashComparison]: {
    label: 'Hash comparison',
    labelPending: 'Comparing hash',
    subSteps: []
  },
  [VerificationSteps.statusCheck]: {
    label: 'Status check',
    labelPending: 'Checking record status',
    subSteps: []
  }
};
exports.language = language;
